pkg/controller/standard.oam.dev/v1alpha1/rollout/handler.go

```diff
@@ -0,0 +1,225 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rollout
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/oam-dev/kubevela/pkg/oam"
+
+	"github.com/oam-dev/kubevela/pkg/controller/core.oam.dev/v1alpha2/applicationrollout"
+
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"k8s.io/klog/v2"
+
+	"github.com/oam-dev/kubevela/apis/standard.oam.dev/v1alpha1"
+
+	v1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/oam-dev/kubevela/pkg/controller/core.oam.dev/v1alpha2/application/assemble"
+	"github.com/oam-dev/kubevela/pkg/oam/util"
+)
+
+type handler struct {
+	*reconciler
+	rollout       *v1alpha1.Rollout
+	sourceRevName string
+	targetRevName string
+	compName      string
+
+	sourceWorkload *unstructured.Unstructured
+	targetWorkload *unstructured.Unstructured
+}
+
+// assembleWorkload help to assemble worklad info, because rollout don't know AppRevision we cannot use assemble func directly.
+// but we do same thing with it in the func
+func (h *handler) assembleWorkload(ctx context.Context) error {
+	workloadOptions := []assemble.WorkloadOption{
+		applicationrollout.RolloutWorkloadName(h.compName),
+		assemble.PrepareWorkloadForRollout(h.compName),
+		applicationrollout.HandleReplicas(ctx, h.compName, h)}
+
+	for _, workloadOption := range workloadOptions {
+		if err := workloadOption.ApplyToWorkload(h.targetWorkload, nil, nil); err != nil {
+			return err
+		}
+		if len(h.sourceRevName) != 0 {
+			if err := workloadOption.ApplyToWorkload(h.sourceWorkload, nil, nil); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// extractWorkloadFromCompRevision extract workload info from component revision
+func (h *handler) extractWorkloadFromCompRevision(ctx context.Context) error {
```

Comment:
this function contains many error that return directly without any logs or errro wrapper. How would you like to debug if it running in production?

Reply:
fiexed